### PR TITLE
Add Toggle Output Scroll Lock action

### PR DESCRIPTION
**Pull request by ArtemGovorov**
As discussed here https://github.com/Microsoft/vscode/pull/18589#issuecomment-273528192.
@isidorn I have implemented the action and added it to the panel, but it's missing the icon. I'm also not sure how to make the action button reflect if the output is locked or not. Perhaps it's possible to use the action `checked` property and style the checked button in the same CSS file where the icon will be referenced, but the issue is that the action is not channel specific (ie there's only one instance of the action per panel, not per channel).

### CODE REVIEW

**Comment by ArtemGovorov** [comment]
>  but the issue is that the action is not channel specific (ie there's only one instance of the action per panel, not per channel).

Ok, so I found how to make the action `checked` state to consider the current active channel, added the commit to support it.

So what's left now is an icon and some CSS for the `action-label` to display the desired checked state (like a border or a separate background).

File: src/vs/workbench/parts/output/common/output.ts [hunk 1]
```diff
@@ -111,6 +111,11 @@ export interface IOutputChannel {
 	output: string;
 
 	/**
+	 * Returns the value indicating whether the channel has scroll locked.
+	 */
+	hasScrollLock: boolean;
```
Comment (isidorn) [inline src/vs/workbench/parts/output/common/output.ts]: I would rename this to `scrollLock`, as `outputChannel.hasScrollLock = false` sounds strange to me

File: src/vs/workbench/parts/output/browser/outputActions.ts [hunk 2]
```diff
@@ -48,6 +49,33 @@ export class ClearOutputAction extends Action {
...
+	private toDispose: IDisposable[] = [];
+
+	constructor(id: string, label: string,
+		@IOutputService private outputService: IOutputService) {
+		super(id, label, 'output-action toggle-output-scroll-lock');
+		this.toDispose.push(this.outputService.onActiveOutputChannel(channel => this._setChecked(this.outputService.getActiveChannel().hasScrollLock)));
+	}
+
+	public run(): TPromise<any> {
+		const activeChannel = this.outputService.getActiveChannel();
```
Comment (isidorn) [inline src/vs/workbench/parts/output/browser/outputActions.ts]: What if there is no active output channel (output hidden) and this command is triggered from the command pallete?
You either need a null check here or even better do not register this action as a global workbench action as I do not see people assigning shortcuts to this action or triggering it via command pallete.

**Comment by isidorn** [comment]
@ArtemGovorov PR looks good overall. Added 2 comments which you can see directly in the commits.
You could play around with the css and just use a temporary fake icon. For example this one
https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/preferences/browser/media/collapseAll.svg#L1

Once @bgashler1 provides a lock icon I can just replace it

**Comment by ArtemGovorov** [comment]
@isidorn Thanks for reviewing it. I have fixed both comments.

**Comment by isidorn** [comment]
@ArtemGovorov great! Let's wait for the icon before we merge this in.
In the meantime feel free to experiment with css and a fake icon as I mentioned above.

**Comment by ArtemGovorov** [comment]
> There seems to be build errors as our travis build is falling, can you also please look into that? As far as I see it seems like the Label and Id are needed on the action.

Oops, my fault. Added these properties back.

> In the meantime feel free to experiment with css and a fake icon as I mentioned above.

Cool, will be doing it now.

**Comment by ArtemGovorov** [comment]
Added the styles for the checked state. As for the icon, foe now I just copy.pasted [the Clear Output icon](https://github.com/Microsoft/vscode/pull/18768/commits/b4ce09aa9e50a02ed21e77574f83fec43d2c658e#diff-c42d4dc78af7b92e87a8e5822dd93ddb) content, so all that's left is to replace the SVG in the 2 icons files (normal and inverted) with the new icon.

ctrb

**Comment by isidorn** [comment]
@ArtemGovorov looks great! Just one more minor thing please revert your changes to `output.contribution.ts` since our hygiene is failing since you introduced an unused import. In the end you should not have any changes in this file - thanks

**Comment by ArtemGovorov** [comment]
@isidorn Thanks, done.

**Comment by bgashler1** [comment]
Here are a couple of different options we could go with.
scroll-lock-icons

And attached are the assets.
[881665_Brad_VSCodeOutputWindowScrollLock_TwoIdeas.zip](https://github.com/Microsoft/vscode/files/724889/881665_Brad_VSCodeOutputWindowScrollLock_TwoIdeas.zip)

**Comment by isidorn** [comment]
Created test plan item https://github.com/Microsoft/vscode/issues/19138
Took the first icon. Will not swtich between icons but will draw a blue border when the action is active as we do in the find widget. I find that less confusing for the user.
Choose the first icon because we never render the scroll in the manner that the second icon is showin, and the first one is more general

**Comment by ArtemGovorov** [comment]
Awesome, thanks! The icon looks great. So excited to see feature is being added to the next release!

**Pull request merged**